File: editor/en/[255]Verify Preorder Sequence in Binary Search Tree.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public boolean verifyPreorder(int[] preorder) {
        Deque<Integer> deque = new ArrayDeque<>();
        int lower = Integer.MIN_VALUE;

        for (int value : preorder) {
            if (value < lower) {
                return false;
            }

            while (!deque.isEmpty() && value > deque.peek()) {
                lower = deque.pop();
            }

            deque.push(value);
        }

        return true;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[255]Verify Preorder Sequence in Binary Search Tree.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        low = float('-inf')
        i = -1
        for p in preorder:
            if p < low:
                return False
            while i >= 0 and p > preorder[i]:
                low = preorder[i]
                i -= 1
            i += 1
            preorder[i] = p

        return True

# leetcode submit region end(Prohibit modification and deletion)
